smart-contracts/contracts/cl-vault/src/test_tube/withdraw_main_pro_rato.rs

```diff
@@ -0,0 +1,401 @@
+        //                     .checked_mul(Decimal256::from_ratio(9_u128, 10_u128))
+        //                     .unwrap(),
+        //             },
+        //         )),
+        //         &[],
+        //         &admin,
+        //     )
+        //     .unwrap();
+
+        // withdraw funds from the main position
+        let token0 = bank
+            .query_balance(&QueryBalanceRequest {
+                address: alice.address(),
+                denom: DENOM_BASE.to_string(),
+            })
+            .unwrap();
+        let token1 = bank
+            .query_balance(&QueryBalanceRequest {
+                address: alice.address(),
+                denom: DENOM_QUOTE.to_string(),
+            })
+            .unwrap();
+        let alice_balance_value_before = get_value_in_asset0(
+            &app,
+            cl_pool_id,
+            token0.balance.unwrap().try_into().unwrap(),
+            token1.balance.unwrap().try_into().unwrap(),
+        )
+        .unwrap();
+
+        let withdraw = wasm
+            .execute(
+                contract_address.as_str(),
+                &ExecuteMsg::Redeem {
+                    recipient: None,
+                    amount: shares_to_withdraw,
+                },
+                &[],
+                &alice,
+            )
+            .unwrap();
+        let source = get_event_attributes_by_ty_and_key(&withdraw, "wasm", vec!["source"]);
+        assert_eq!(source[0].value, "all_positions");
+
+        let token0 = bank
+            .query_balance(&QueryBalanceRequest {
+                address: alice.address(),
+                denom: DENOM_BASE.to_string(),
+            })
+            .unwrap();
+        let token1 = bank
+            .query_balance(&QueryBalanceRequest {
+                address: alice.address(),
+                denom: DENOM_QUOTE.to_string(),
+            })
+            .unwrap();
+        let alice_balance_value_after = get_value_in_asset0(
+            &app,
+            cl_pool_id,
+            token0.balance.unwrap().try_into().unwrap(),
+            token1.balance.unwrap().try_into().unwrap(),
+        )
+        .unwrap();
+
+        let second_difference = alice_balance_value_after - alice_balance_value_before;
+
+        let positions: PositionsResponse = wasm
+            .query(
+                contract_address.as_str(),
+                &QueryMsg::VaultExtension(ExtensionQueryMsg::ConcentratedLiquidity(
+                    ClQueryMsg::Positions {},
+                )),
+            )
+            .unwrap();
+        // since we are withdrawing from multiple different positions here in the second difference, we might encounter multiple
+        // moments where we round down, hence we can be off by 1 * the amount of positions + 1 for the free funds.
+        // since each of those calculations is a potential round down
+        let allowed_absolute_diff = positions.positions.len() as u128 + 1_u128;
+        assert_eq_with_diff!(
```

Comment:
why not exact comparison?

Reply:
Everytime we calculate a percentage, we need to round down to make sure that the vault never "drains" itself (which rounding up will do). If we have 5 positions, we'd round down 5 times for the positions and 1 time for the free balance, because of which the user would get 5 less microtokens(uosmo etc). On the main position withdraw, the user would get 1 rounding down from the position. The tests needs to account for that difference

Reply:
This can be rewritten to assert the `allowed_absolute_diff`, will change this